prov/sockets/src/sock_cntr.c

```diff
@@ -301,6 +304,10 @@ static int sock_cntr_wait(struct fid_cntr *cntr, uint64_t threshold,
 		ret = fi_wait_cond(&_cntr->cond, &_cntr->mut, timeout);
 	}
 
+	if (atomic_get(&_cntr->value) >= threshold &&
+	    atomic_get(&_cntr->last_read_val) < threshold) {
```

Comment:
Yeah, I had thought about this. I went this way because what if some other thread does an fi_cntr_read() and got a higher value than threshold. So, in that case last-read-val will be set to something less than what application already read.

But after taking a closer look, even this can run into the same issue. (cntr_read() can happen after the if condition and before the atomic_set.)